augur/io/sqlite3.py

```diff
@@ -0,0 +1,137 @@
+
+
+class Sqlite3Database:
+    def __init__(self, file:str=''):
+
+        self.file = file
+        """Database file. Empty string if using an in-memory database."""
+
+        self.use_in_memory_db = not self.file
+        """Whether to use an in-memory database."""
+
+        if self.use_in_memory_db:
+            # More info: https://www.sqlite.org/inmemorydb.html
+            self.in_memory_db_connection = sqlite3.connect(":memory:")
+            """Persistent connection to an in-memory database."""
+
+    def connect(self, **connect_kwargs) -> sqlite3.Connection:
+        """Return a connection to the SQLite database.
+        If using an in-memory database, the same connection is returned for all calls to this method.
+        Otherwise, a new connection is returned every time."""
+        if self.use_in_memory_db:
+            return self.in_memory_db_connection
+        return sqlite3.connect(self.file, **connect_kwargs)
+
+    def execute(self, *args):
+        """Execute under a connection context."""
+        with self.connect() as con:
+            con.execute(*args)
+
+    def executemany(self, *args):
+        """Executemany under a connection context."""
+        # TODO: benchmark performance with cursor.executemany / con.commit at the end
+        with self.connect() as con:
+            con.executemany(*args)
+
+    def create_unique_index(self, table_name:str, column:str):
+        """Create a unique index on a column in a table."""
+        index_name = f'index_{table_name}_{column}'
+        self.execute(f"""
+            CREATE UNIQUE INDEX {sanitize_identifier(index_name)}
+            ON {sanitize_identifier(table_name)} ({sanitize_identifier(column)})
+        """)
+
+    def query_to_file(self, query:str, path:str, header=True, chunksize:int=10000, columns_to_exclude:List[str]=[], **to_csv_kwargs):
+        """Query the database and write results to a tabular file.
+
+        Parameters
+        ----------
+        query
+            SQLite query string.
+        path
+            Path to the output file.
+        header
+            Write out the column names.
+        chunksize
+            Maximum number of rows to load into memory at once.
+        columns_to_exclude
+            Column names to exclude from output.
+        to_csv_kwargs
+            Additional keyword arguments passed to pandas.to_csv().
+        """
+        with self.connect() as con:
+            # Get results in chunks so they are not loaded into memory all at once.
+            df_chunks = pd.read_sql_query(query, con, chunksize=chunksize)
+
+            # Overwrite the file if it exists.
+            mode = 'w'
+
+            for df_chunk in df_chunks:
+                # There is no way to exclude columns in a SQLite query, so it is implemented as an option here.
+                if columns_to_exclude:
+                    for col in columns_to_exclude:
+                        df_chunk.drop(col, axis=1, inplace=True)
```

Comment:
What if this logic is handled in the query instead of at the pandas level? For example, why not `SELECT {",".join(columns_to_keep)} from {table_name}` instead of `SELECT * FROM {table_name}`?

Reply:
This is also an artifact of the other PR. The feature is unused here, so I've removed it: 963bbe6ef8fb7dd3490f32c591d1c528e0ddf2b9

For context, in #854 it is used here: https://github.com/nextstrain/augur/blob/0385409cf6eeddf91d4c95551e73e3251a1fdd22/augur/filter_support/db/sqlite.py#L789-L808

I didn't handle it on the query level because the column names are not stored anywhere besides the database. I suppose this could have worked:

```py
cursor = connection.execute(f'SELECT * FROM {table_name}')
columns = cursor.fetchone().keys()
columns_to_keep = columns.remove(ROW_ORDER_COLUMN)
query = f'SELECT {",".join(columns_to_keep)} FROM {table_name}'
…
```

But in any case, that's something that can be discussed later once the feature is re-added with actual usage.

Reply:
(You can also introspect the schema without selecting a data row.)